File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.Example1/LoggingProxy.cs
namespace ProxyDesignPattern.Example1;

/// <summary>
/// LOGGING PROXY - Logs all video service operations for analytics.
/// </summary>
public class LoggingProxy : IVideoService
{
    private readonly IVideoService _innerService;
    private readonly List<string> _logs = new();

    public LoggingProxy(IVideoService innerService)
    {
        _innerService = innerService;
    }

    public Video? GetVideoInfo(string videoId)
    {
        var timestamp = DateTime.Now.ToString("HH:mm:ss");
        _logs.Add($"[{timestamp}] GetVideoInfo called for: {videoId}");
        Console.WriteLine($"[LoggingProxy] Logging: GetVideoInfo({videoId})");
        
        return _innerService.GetVideoInfo(videoId);
    }

    public void StreamVideo(string videoId)
    {
        var timestamp = DateTime.Now.ToString("HH:mm:ss");
        _logs.Add($"[{timestamp}] StreamVideo called for: {videoId}");
        Console.WriteLine($"[LoggingProxy] Logging: StreamVideo({videoId})");
        
        _innerService.StreamVideo(videoId);
    }

    public void PrintLogs()
    {
        Console.WriteLine("\n--- Activity Logs ---");
        foreach (var log in _logs)
        {
            Console.WriteLine(log);
        }
        Console.WriteLine("---------------------");
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.Example1/src/LoggingProxy.ts
import { IVideoService } from './IVideoService';
import { Video } from './Models';

// LOGGING PROXY — logs all operations
export class LoggingProxy implements IVideoService {
  private readonly logs: string[] = [];

  constructor(private readonly innerService: IVideoService) {}

  getVideoInfo(videoId: string): Video | null {
    const ts = new Date().toLocaleTimeString();
    this.logs.push(`[${ts}] GetVideoInfo called for: ${videoId}`);
    console.log(`[LoggingProxy] Logging: getVideoInfo(${videoId})`);
    return this.innerService.getVideoInfo(videoId);
  }

  streamVideo(videoId: string): void {
    const ts = new Date().toLocaleTimeString();
    this.logs.push(`[${ts}] StreamVideo called for: ${videoId}`);
    console.log(`[LoggingProxy] Logging: streamVideo(${videoId})`);
    this.innerService.streamVideo(videoId);
  }

  printLogs(): void {
    console.log('\n--- Activity Logs ---');
    for (const log of this.logs) console.log(log);
    console.log('---------------------');
  }
}
